services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/KafkaBasedCommandConsumerFactoryImpl.java

```diff
@@ -92,12 +94,14 @@ public KafkaBasedCommandConsumerFactoryImpl(
 
         final KafkaBasedInternalCommandSender internalCommandSender = new KafkaBasedInternalCommandSender(
                 kafkaProducerFactory, kafkaProducerConfig, tracer);
-        commandHandler = new KafkaBasedMappingAndDelegatingCommandHandler(tenantClient, commandTargetMapper,
-                internalCommandSender, tracer);
+        commandHandler = new KafkaBasedMappingAndDelegatingCommandHandler(tenantClient, commandQueue,
+                commandTargetMapper, internalCommandSender, tracer);
         final Map<String, String> consumerConfig = kafkaConsumerConfig.getConsumerConfig("consumer");
         consumerConfig.put(ConsumerConfig.GROUP_ID_CONFIG, "cmd-router-group");
-        kafkaConsumer = new HonoKafkaConsumer(vertx, COMMANDS_TOPIC_PATTERN,
+        kafkaConsumer = new AsyncHandlingAutoCommitKafkaConsumer(vertx, COMMANDS_TOPIC_PATTERN,
                 commandHandler::mapAndDelegateIncomingCommandMessage, consumerConfig);
+        kafkaConsumer.setOnPartitionsAssignedHandler(
```

Comment:
I don't think intervening with the command processing via extra synchronization ("don't block the event loop") or waiting for some Futures added in between would be a good idea. And when "kafkaProducer.send()" has been invoked to forward the command on the internal command topic, there is no stopping that anyway.

> During re-balancing, the commands are still processed and consumers try to commit them too. Won't those commit fails during re-balancing?

During the rebalance, happening on the Kafka polling thread, the commands are indeed still processed on the event loop thread, but they are not committed there. The commands are just marked as "completed". The next commit happening on the Kafka polling thread will then take these completed commands into account when determining the offsets to be committed.

The scenario of a partition getting revoked (and afterwards not getting re-assigned) while a command is in the process of getting published on the internal command topic, might happen. But I think it's a narrow timeframe in which that can happen. And in case a command gets sent twice in such a case (the 2nd time by the consumer that the partition assignment has switched to), the duplicate command will indeed be ignored by the target protocol adapter (via #2596). So I don't think we should add further synchronization/locks in the command router for that.